src/Functions/CustomWeekTransforms.h

```diff
@@ -110,6 +86,30 @@ struct ToStartOfWeekImpl
     using FactorTransform = ZeroTransform;
 };
 
+struct ToWeekImpl
+{
+    static constexpr auto name = "toWeek";
+
+    static inline UInt8 execute(Int64 t, UInt8 week_mode, const DateLUTImpl & time_zone)
+    {
+        // TODO: ditch conversion to DayNum, since it doesn't support extended range.
+        YearWeek yw = time_zone.toYearWeek(time_zone.toDayNum(t), week_mode);
+        return yw.second;
+    }
+    static inline UInt8 execute(UInt32 t, UInt8 week_mode, const DateLUTImpl & time_zone)
+    {
+        YearWeek yw = time_zone.toYearWeek(time_zone.toDayNum(t), week_mode);
+        return yw.second;
+    }
+    static inline UInt8 execute(UInt16 d, UInt8 week_mode, const DateLUTImpl & time_zone)
+    {
+        YearWeek yw = time_zone.toYearWeek(DayNum(d), week_mode);
+        return yw.second;
+    }
+
+    using FactorTransform = ToStartOfWeekImpl;
```

Comment:
`ToStartOfYearImpl`? We need to check, that two dates belong to the same year.

Reply:
but `toStartOfWeek` contains year
```
┌─toStartOfWeek(now())─┐
│           2021-05-23 │
└──────────────────────┘
```

Reply:
Yes, but `toWeek` in range of one year can be treated as monotonic, but `toStartOfWeek` don't allow this.

Reply:
Right.